plugins/search-backend/src/service/AuthorizedSearchEngine.test.ts

```diff
@@ -0,0 +1,382 @@
+    );
+    expect(mockedAuthorize).toHaveBeenCalledTimes(1);
+    expect(mockedAuthorize).toHaveBeenLastCalledWith(
+      [
+        { permission: defaultTypes[typeUsers].visibilityPermission },
+        { permission: defaultTypes[typeTemplates].visibilityPermission },
+      ],
+      { token: 'token' },
+    );
+  });
+
+  it('should skip sending request for types that are not allowed', async () => {
+    mockedAuthorize.mockImplementation(async queries => {
+      return queries.map(query => {
+        if (
+          query.permission.name ===
+          defaultTypes.users.visibilityPermission?.name
+        ) {
+          return {
+            result: AuthorizeResult.DENY,
+          };
+        }
+        return {
+          result: AuthorizeResult.ALLOW,
+        };
+      });
+    });
+
+    await authorizedSearchEngine.query({ term: '' }, options);
+
+    expect(mockedQuery).toHaveBeenCalledWith(
+      { term: '', types: ['templates'] },
+      { token: 'token' },
+    );
+
+    expect(mockedAuthorize).toHaveBeenCalledTimes(1);
+  });
+
+  it('should perform result-by-result filtering', async () => {
+    const usersWithAuth = generateSampleResults(typeUsers, true);
+    const templatesWithAuth = generateSampleResults(typeTemplates, true);
+
+    const resultsWithAuth = usersWithAuth.concat(templatesWithAuth);
+
+    mockedQuery.mockImplementation(async () => ({
+      results: resultsWithAuth,
+    }));
+
+    const userToBeReturned = 8;
+
+    // TODO(vinzscam): I am not sure if such authorizer makes sense
+    mockedAuthorize.mockImplementation(async queries =>
+      queries.map(query => {
+        if (
+          query.permission.name ===
+          defaultTypes.users.visibilityPermission?.name
+        ) {
+          if (query.resourceRef) {
+            return {
+              result: query.resourceRef.endsWith(userToBeReturned.toString())
+                ? AuthorizeResult.ALLOW
+                : AuthorizeResult.DENY,
+            };
+          }
+          return {
+            result: AuthorizeResult.CONDITIONAL,
+            // TODO(vinzscam): is something like this allowed?
+            // I guess NO, but I am not sure.
+            // if yes, who is responsible for evaluating this?
+            conditions: { not: { rule: 'ok', params: [] } },
+          };
+        }
+
+        return {
+          result: AuthorizeResult.DENY,
+        };
+      }),
+    );
+
+    await expect(
+      authorizedSearchEngine.query({ term: '' }, options),
+    ).resolves.toEqual({ results: [usersWithAuth[userToBeReturned]] });
+
+    expect(mockedQuery).toHaveBeenCalledWith(
+      { term: '', types: ['users'] },
+      { token: 'token' },
+    );
+  });
+
+  it('should perform search until the target number of results is reached', async () => {
+    mockedAuthorize.mockImplementation(async queries =>
+      queries.map(query => {
+        if (query.resourceRef) {
+          return { result: AuthorizeResult.ALLOW };
```

Comment:
Along with this test, I think it would also be good to test the case where some results are visible and some are hidden - right now we are testing the "repagination" performed by the AuthorizedSearchEngine, but I don't believe we're testing the filtering aspect, because all results are visible.